File: appanalytics/ios/Classes/TalkingDataAppAnalyticsPlugin.m
#import "TalkingDataAppAnalyticsPlugin.h"
#import "TalkingData.h"

@implementation TalkingDataAppAnalyticsPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"TalkingData_AppAnalytics"
            binaryMessenger:[registrar messenger]];
    TalkingDataAppAnalyticsPlugin* instance = [[TalkingDataAppAnalyticsPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}


+(void)pluginSessionStart:(NSString*)session withChannelId:(NSString*)channelID
{
    [TalkingData sessionStarted:session withChannelId:channelID];
}


-(id)checkArgument:(NSDictionary*)argument forKey:(NSString*)key ofType:(Class)clazz
{
    if (key == nil || argument == nil || clazz == nil  ) {
        return nil;
    }
    id arg = [argument objectForKey:key];
    if (arg == nil) {
        return nil;
    }
    if (![arg isKindOfClass:clazz]) {
        return nil;
    }
    return arg;
}


-(TDAccountType)accountTypeConvert:(NSString*)accTypeStr
{
    if ([accTypeStr isEqualToString:@"ANONYMOUS"]) {
        return TDAccountTypeAnonymous;
    }else if ([accTypeStr isEqualToString:@"REGISTERED"]){
        return TDAccountTypeRegistered;
    }else if ([accTypeStr isEqualToString:@"SINA_WEIBO"]){
        return TDAccountTypeSinaWeibo;
    }else if ([accTypeStr isEqualToString:@"QQ"]){
        return TDAccountTypeQQ;
    }else if ([accTypeStr isEqualToString:@"QQ_WEIBO"]){
        return TDAccountTypeTencentWeibo;
    }else if ([accTypeStr isEqualToString:@"ND91"]){
        return TDAccountTypeND91;
    }else if ([accTypeStr isEqualToString:@"WEIXIN"]){
        return TDAccountTypeWeiXin;
    }else if ([accTypeStr isEqualToString:@"TYPE1"]){
        return TDAccountTypeType1;
    }else if ([accTypeStr isEqualToString:@"TYPE2"]){
        return TDAccountTypeType2;
    }else if ([accTypeStr isEqualToString:@"TYPE3"]){
        return TDAccountTypeType3;
    }else if ([accTypeStr isEqualToString:@"TYPE4"]){
        return TDAccountTypeType4;
    }else if ([accTypeStr isEqualToString:@"TYPE5"]){
        return TDAccountTypeType5;
    }else if ([accTypeStr isEqualToString:@"TYPE6"]){
        return TDAccountTypeType6;
    }else if ([accTypeStr isEqualToString:@"TYPE7"]){
        return TDAccountTypeType7;
    }else if ([accTypeStr isEqualToString:@"TYPE8"]){
        return TDAccountTypeType8;
    }else if ([accTypeStr isEqualToString:@"TYPE9"]){
        return TDAccountTypeType9;
    }else if ([accTypeStr isEqualToString:@"TYPE10"]){
        return TDAccountTypeType10;
    }
    return TDAccountTypeAnonymous;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"onPageStart" isEqualToString:call.method]){
        NSString * pageName = [self checkArgument:call.arguments forKey:@"pageName" ofType:[NSString class]];
        if (pageName) {
            [TalkingData trackPageBegin:pageName];
        }
    } else if ([@"onPageEnd" isEqualToString:call.method]){
        NSString * pageName = [self checkArgument:call.arguments forKey:@"pageName" ofType:[NSString class]];
        if (pageName) {
            [TalkingData trackPageEnd:pageName];
        }
    } else  if ([@"onRegister" isEqualToString:call.method]){
        NSString * accountID = [self checkArgument:call.arguments forKey:@"accountID" ofType:[NSString class]];
        NSString * accountType = [self checkArgument:call.arguments forKey:@"accountType" ofType:[NSString class]];
        NSString * name = [self checkArgument:call.arguments forKey:@"name" ofType:[NSString class]];
        if (accountID && accountType && name) {
            TDAccountType acctype = [self accountTypeConvert:accountType];
            [TalkingData onRegister:accountID type:acctype name:name];
        }
    } else if ([@"onLogin" isEqualToString:call.method]){
        NSString * accountID = [self checkArgument:call.arguments forKey:@"accountID" ofType:[NSString class]];
        NSString * accountType = [self checkArgument:call.arguments forKey:@"accountType" ofType:[NSString class]];
        NSString * name = [self checkArgument:call.arguments forKey:@"name" ofType:[NSString class]];
        if (accountID && accountType && name) {
            TDAccountType acctype = [self accountTypeConvert:accountType];
            [TalkingData onLogin:accountID type:acctype name:name];
        }
    } else if ([@"onEvent" isEqualToString:call.method]){
        NSString * eventID = [self checkArgument:call.arguments forKey:@"eventID" ofType:[NSString class]];
        NSString * eventLabel = [self checkArgument:call.arguments forKey:@"eventLabel" ofType:[NSString class]];
        NSDictionary * params = [self checkArgument:call.arguments forKey:@"params" ofType:[NSDictionary class]];
        if (eventID) {
            [TalkingData trackEvent:eventID label:eventLabel parameters:params];
        }
    } else if ([@"onViewItem" isEqualToString:call.method]){
        NSString* category = [self checkArgument:call.arguments forKey:@"category" ofType:[NSString class]];
        NSString* itemID = [self checkArgument:call.arguments forKey:@"itemID" ofType:[NSString class]];
        NSString* name = [self checkArgument:call.arguments forKey:@"name" ofType:[NSString class]];
        NSNumber* unitPrice = [self checkArgument:call.arguments forKey:@"unitPrice" ofType:[NSNumber class]];
        [TalkingData onViewItem:itemID category:category name:name unitPrice:unitPrice.intValue];
        
    } else if ([@"onAddItemToShoppingCart" isEqualToString:call.method]){
        NSNumber * amount = [self checkArgument:call.arguments forKey:@"amount" ofType:[NSNumber class]];
        NSString * category = [self checkArgument:call.arguments forKey:@"category" ofType:[NSString class]];
        NSString * itemID = [self checkArgument:call.arguments forKey:@"itemID" ofType:[NSString class]];
        NSString * name = [self checkArgument:call.arguments forKey:@"name" ofType:[NSString class]];
        NSNumber * uniprice = [self checkArgument:call.arguments forKey:@"unitPrice" ofType:[NSNumber class]];
        [TalkingData onAddItemToShoppingCart:itemID category:category name:name unitPrice:uniprice.intValue amount:amount.intValue];
    } else if ([@"onViewShoppingCart" isEqualToString:call.method]){
        NSArray * shoppingCartDetails = [self checkArgument:call.arguments forKey:@"shoppingCartDetails" ofType:[NSArray class]];
        TalkingDataShoppingCart * sc = [TalkingDataShoppingCart createShoppingCart];
        for (NSDictionary* each in shoppingCartDetails) {
            NSNumber * amount = [self checkArgument:each forKey:@"amount" ofType:[NSNumber class]];
            NSString * category = [self checkArgument:each forKey:@"category" ofType:[NSString class]];
            NSString * itemID = [self checkArgument:each forKey:@"itemID" ofType:[NSString class]];
            NSString * name = [self checkArgument:each forKey:@"name" ofType:[NSString class]];
            NSNumber * uniprice = [self checkArgument:each forKey:@"unitPrice" ofType:[NSNumber class]];
            [sc addItem:itemID category:category name:name unitPrice:uniprice.intValue amount:amount.intValue];
        }
        [TalkingData onViewShoppingCart:sc];
    } else if ([@"onPlaceOrder" isEqualToString:call.method]){
        NSString * accountID = [self checkArgument:call.arguments forKey:@"accountID" ofType:[NSString class]];
        NSString * currencyType = [self checkArgument:call.arguments forKey:@"currencyType" ofType:[NSString class]];
        NSArray * orderDetails = [self checkArgument:call.arguments forKey:@"orderDetails" ofType:[NSArray class]];
        NSString * orderID = [self checkArgument:call.arguments forKey:@"orderID" ofType:[NSString class]];
        NSNumber * totalPrice = [self checkArgument:call.arguments forKey:@"totalPrice" ofType:[NSNumber class]];
        
        TalkingDataOrder * order = [TalkingDataOrder createOrder:orderID total:totalPrice.intValue currencyType:currencyType];
        
        for (NSDictionary * each in orderDetails) {
            NSNumber * amount = [self checkArgument:each forKey:@"amount" ofType:[NSNumber class]];
            NSString * category = [self checkArgument:each forKey:@"category" ofType:[NSString class]];
            NSString * itemID = [self checkArgument:each forKey:@"itemID" ofType:[NSString class]];
            NSString * name = [self checkArgument:each forKey:@"name" ofType:[NSString class]];
            NSNumber * uniprice = [self checkArgument:each forKey:@"unitPrice" ofType:[NSNumber class]];
            [order addItem:itemID category:category name:name unitPrice:uniprice.intValue amount:amount.intValue];
        }
        [TalkingData onPlaceOrder:accountID order:order];
        
    } else if ([@"onOrderPaySucc" isEqualToString:call.method]){
        NSString * accountID = [self checkArgument:call.arguments forKey:@"accountID" ofType:[NSString class]];
        NSString * currencyType = [self checkArgument:call.arguments forKey:@"currencyType" ofType:[NSString class]];
        NSArray * orderDetails = [self checkArgument:call.arguments forKey:@"orderDetails" ofType:[NSArray class]];
        NSString * orderID = [self checkArgument:call.arguments forKey:@"orderID" ofType:[NSString class]];
        NSString * payType = [self checkArgument:call.arguments forKey:@"payType" ofType:[NSString class]];
        NSNumber * totalPrice = [self checkArgument:call.arguments forKey:@"totalPrice" ofType:[NSNumber class]];
        
        TalkingDataOrder * order = [TalkingDataOrder createOrder:orderID total:totalPrice.intValue currencyType:currencyType];
        
        for (NSDictionary * each in orderDetails) {
            NSNumber * amount = [self checkArgument:each forKey:@"amount" ofType:[NSNumber class]];
            NSString * category = [self checkArgument:each forKey:@"category" ofType:[NSString class]];
            NSString * itemID = [self checkArgument:each forKey:@"itemID" ofType:[NSString class]];
            NSString * name = [self checkArgument:each forKey:@"name" ofType:[NSString class]];
            NSNumber * uniprice = [self checkArgument:each forKey:@"unitPrice" ofType:[NSNumber class]];
            [order addItem:itemID category:category name:name unitPrice:uniprice.intValue amount:amount.intValue];
        }
        [TalkingData onOrderPaySucc:accountID payType:payType order:order];
    }else if ([@"setGlobalKV" isEqualToString:call.method]) {
        NSString * key = [self checkArgument:call.arguments forKey:@"key" ofType:[NSString class]];
        id value = call.arguments[@"value"];
        if (value) {
            [TalkingData setGlobalKV:key value:value];
        }
    }else if ([@"removeGlobalKV" isEqualToString:call.method]){
        NSString * key = [self checkArgument:call.arguments forKey:@"key" ofType:[NSString class]];
        [TalkingData removeGlobalKV:key];
    }else if ([@"getDeviceID" isEqualToString:call.method]){
        NSString * deviceid = [TalkingData getDeviceID];
        result(deviceid);
    } else {
        result(FlutterMethodNotImplemented);
    }
}


@end

File: appanalytics/android/src/main/java/com/talkingdata/plugin/appanalytics/TalkingDataAppAnalyticsPlugin.java
package com.talkingdata.plugin.appanalytics;

import android.content.Context;

import com.tendcloud.tenddata.Order;
import com.tendcloud.tenddata.ShoppingCart;
import com.tendcloud.tenddata.TCAgent;
import com.tendcloud.tenddata.TDAccount;

import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** TalkingDataAppAnalyticsPlugin */
public class TalkingDataAppAnalyticsPlugin implements MethodCallHandler {
  private Context context;

  private TalkingDataAppAnalyticsPlugin(Context context){
    this.context = context;
  }

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "TalkingData_AppAnalytics");
    channel.setMethodCallHandler(new TalkingDataAppAnalyticsPlugin(registrar.context().getApplicationContext()));
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch (call.method){
      case "getDeviceID":
        result.success(TCAgent.getDeviceId(context));
        break;
      case "onPageStart":
        TCAgent.onPageStart(context, (String) call.argument("pageName"));
        break;
      case "onPageEnd":
        TCAgent.onPageEnd(context, (String) call.argument("pageName"));
        break;
      case "onEvent":
        String eventID = call.argument("eventID");
        String eventLabel = call.argument("eventLabel");
        Map params = null;
        if (call.argument("params") instanceof Map){
          params = call.argument("params");
        }
        TCAgent.onEvent(context, eventID, eventLabel, params);
        break;
      case "setGlobalKV":
        String globalKey = call.argument("key");
        Object globalValue = call.argument("value");
        TCAgent.setGlobalKV(globalKey, globalValue);
        break;
      case "removeGlobalKV":
        String key = call.argument("key");
        TCAgent.removeGlobalKV(key);
        break;
      case "onRegister":
        String accountID = call.argument("accountID");
        String accountType = call.argument("accountType");
        String name = call.argument("name");
        TCAgent.onRegister(accountID, TDAccount.AccountType.valueOf(accountType), name);
        break;
      case "onLogin":
        accountID = call.argument("accountID");
        accountType = call.argument("accountType");
        name = call.argument("name");
        TCAgent.onLogin(accountID, TDAccount.AccountType.valueOf(accountType), name);
        break;
      case "onPlaceOrder":
        TCAgent.onPlaceOrder(
                (String) call.argument("accountID"),
                getOrderFromFlutter(call)
        );
        break;
      case "onViewShoppingCart":
        ShoppingCart shoppingCart = ShoppingCart.createShoppingCart();
        List<Map> shoppingCartDetails = call.argument("shoppingCartDetails");
        for (int i = 0; i < shoppingCartDetails.size(); i++){
          Map map = shoppingCartDetails.get(i);
          shoppingCart.addItem(
                  (String)map.get("itemID"),
                  (String)map.get("category"),
                  (String)map.get("name"),
                  (int)map.get("unitPrice"),
                  (int)map.get("amount")
          );
        }

        TCAgent.onViewShoppingCart(shoppingCart);
        break;
      case "onAddItemToShoppingCart":
        TCAgent.onAddItemToShoppingCart(
                (String) call.argument("itemID"),
                (String) call.argument("category"),
                (String) call.argument("name"),
                (int) call.argument("unitPrice"),
                (int) call.argument("amount")
        );
        break;
      case "onOrderPaySucc":
        TCAgent.onOrderPaySucc(
                (String) call.argument("accountID"),
                (String) call.argument("payType"),
                getOrderFromFlutter(call)
        );
        break;
      case "onViewItem":
        TCAgent.onViewItem(
                (String) call.argument("itemID"),
                (String) call.argument("category"),
                (String) call.argument("name"),
                (int) call.argument("unitPrice")
        );
        break;
      default:
        result.notImplemented();
        break;
    }
  }


  private Order getOrderFromFlutter(MethodCall call){
    Order order = null;
    try{
      String orderID = call.argument("orderID");
      int totalPrice = call.argument("totalPrice");
      String currencyType = call.argument("currencyType");

      List orderDetails = call.argument("orderDetails");

      order = Order.createOrder(orderID, totalPrice, currencyType);
      for (int i = 0; i < orderDetails.size(); i++){
        Map<String, Object> map = (Map) orderDetails.get(i);
        String id = String.valueOf(map.get("id"));
        String category = String.valueOf(map.get("category"));
        String name = String.valueOf(map.get("name"));
        int unitPrice = (int) map.get("unitPrice");
        int amount = (int) map.get("amount");
        order.addItem(id, category, name, unitPrice, amount);
      }
    }catch (Throwable t){
      t.printStackTrace();
    }

    return order;
  }
}
